### PR TITLE
EPMDJ-9661 Optimize collections storing

### PR DESCRIPTION
**Pull request by ViktoriiaLebedeva**

### CODE REVIEW

File: src/main/kotlin/Core.kt [hunk 1]
```diff
@@ -160,17 +170,37 @@ suspend inline fun <reified T : Any> Transaction.findById(
...
+): Iterable<T> = transaction {
+    val entities: MutableList<T> = mutableListOf()
+    val idString = ids.joinToString { "'$it'" }
+    val stm = "select JSON_BODY FROM $tableName WHERE ID in ($idString)"
+    val statement = (connection.connection as HikariProxyConnection).createStatement()
+    statement.fetchSize = FETCH_AND_PUSH_LIMIT
+    statement.executeQuery(stm).let { rs ->
+        while (rs.next()) {
+            val fff = json.decodeFromStream(elementSerializer, rs.getBinaryStream(1))
+            entities.add(fff)
```
Comment (Mazel-Tovr) [inline src/main/kotlin/Core.kt:198]:
```suggestion
            entities.add(json.decodeFromStream(elementSerializer, rs.getBinaryStream(1)))
```

File: src/main/kotlin/Core.kt [hunk 2]
```diff
@@ -214,30 +244,88 @@ inline fun <reified T : Any> Transaction.storeAsString(
...
+        }
+    } finally {
+        file.delete()
+    }
+}
+
+fun readerToString(value: Reader, maxLength: Int): String? {
+    return try {
+        val bufferSize = min(maxLength, 1024)
+        val v = StringBuilder(bufferSize)
```
Comment (Mazel-Tovr) [inline src/main/kotlin/Core.kt:300]:
```suggestion
        val result = StringBuilder(bufferSize)
```

File: src/main/kotlin/Utils.kt [hunk 3]
```diff
@@ -15,27 +15,40 @@
...
+fun Any.encodeId(): String = when (this) {
+    is String -> this
+    is Enum<*> -> toString()
+    else -> json.encodeToString(unchecked(this::class.serializer()), this)
+}
 
-fun Any.encodeId(): String = this as? String ?: (if(this is Enum<*>) this.toString() else null)?: json.encodeToString(unchecked(this::class.serializer()), this)
+fun <T : Any> KClass<T>.customSerializer(classLoader: ClassLoader): KSerializer<T> {
+    return CustomSerializer(this.serializer(), classLoader)
+}
```
Comment (Mazel-Tovr) [inline src/main/kotlin/Utils.kt:51]:
```suggestion
fun <T : Any> KClass<T>.customSerializer(
classLoader: ClassLoader,
): KSerializer<T>  = CustomSerializer(this.serializer(), classLoader)

```

**Review by Mazel-Tovr — CHANGES_REQUESTED** [review]

File: src/main/kotlin/Core.kt [hunk 4]
```diff
@@ -214,30 +244,88 @@ inline fun <reified T : Any> Transaction.storeAsString(
...
     stm.executeUpdate()
 }
 
+inline fun <reified T : Any> storeCollection(
+    tableName: String,
+    any: Iterable<T>,
+    parentId: Int?,
+    elementSerializer: KSerializer<T>
+) {
+    val file = File.createTempFile("prefix-", "-postfix")
```
Comment (ViktoriiaLebedeva) [inline src/main/kotlin/Core.kt:265]: add // TODO EPMDJ-9370 Remove file creating

File: src/main/kotlin/serializer/Serializers.kt [hunk 5]
```diff
@@ -67,3 +53,136 @@ object BinarySerializer : KSerializer<ByteArray> {
     override val descriptor: SerialDescriptor
         get() = buildClassSerialDescriptor("Binary")
 }
+
+/**
+ * This serializer stores Collections, in the future Map and ByteArray in a separate table
+ * Hashcode of the collection item is used as the id.
```
Comment (ViktoriiaLebedeva) [inline src/main/kotlin/serializer/Serializers.kt:59]: fix: id of parent

File: src/main/kotlin/serializer/Serializers.kt [hunk 6]
```diff
@@ -67,3 +53,136 @@ object BinarySerializer : KSerializer<ByteArray> {
...
+ * @param classLoader is necessary for deserializing the class
+ */
+class CustomSerializer<T>(
```
Comment (ViktoriiaLebedeva) [inline src/main/kotlin/serializer/Serializers.kt:62]: Rename - DsmSerializer

File: src/main/kotlin/serializer/Serializers.kt [hunk 7]
```diff
@@ -67,3 +53,136 @@ object BinarySerializer : KSerializer<ByteArray> {
...
+                deserializer: DeserializationStrategy<T>,
+                previousValue: T?
+            ): T {
+                return when (deserializer) {
+                    ByteArraySerializer(), is MapLikeSerializer<*, *, *, *> -> {
+                        compositeDecoder.decodeSerializableElement(descriptor, index, deserializer)
+                    }
+                    is AbstractCollectionSerializer<*, *, *> -> {
+                        val elementDescriptor = deserializer.descriptor.getElementDescriptor(0)
+                        elementDescriptor.takeIf { it.kind !is PrimitiveKind }?.run {
```
Comment (ViktoriiaLebedeva) [inline src/main/kotlin/serializer/Serializers.kt:150]: if-else

File: src/main/kotlin/serializer/Serializers.kt [hunk 8]
```diff
@@ -67,3 +53,136 @@ object BinarySerializer : KSerializer<ByteArray> {
...
+                            val kClass = classLoader.loadClass(serialName).kotlin
```
Comment (ViktoriiaLebedeva) [inline src/main/kotlin/serializer/Serializers.kt:151]: Rename - elementClass

File: src/main/kotlin/Core.kt [hunk 9]
```diff
@@ -160,17 +170,37 @@ suspend inline fun <reified T : Any> Transaction.findById(
...
         }
     }
     entity
 }
 
+inline fun <reified T : Any> findByIds(
+    tableName: String,
+    ids: Collection<T>,
+    elementSerializer: KSerializer<T>
+): Iterable<T> = transaction {
```
Comment (ViktoriiaLebedeva) [inline src/main/kotlin/Core.kt:189]: if ids is empty return empty collection

File: src/test/kotlin/TestClasses.kt [hunk 10]
```diff
@@ -87,6 +87,7 @@ data class PayloadWithList(
 @Serializable
 data class ObjectWithSetField(
     @Id val id: String,
+    //todo does't work
```
Comment (ViktoriiaLebedeva) [inline src/test/kotlin/TestClasses.kt:90]: remove

File: src/main/kotlin/Core.kt [hunk 11]
```diff
@@ -160,17 +170,42 @@ suspend inline fun <reified T : Any> Transaction.findById(
...
     entity
 }
 
+inline fun <reified T : Any> findByIds(
+    ids: Collection<T>,
+    elementClass: KClass<*>,
+    elementSerializer: KSerializer<T>
+): Iterable<T> = transaction {
+    val entities: MutableList<T> = mutableListOf()
+    if (ids.isEmpty()) return@transaction entities
```
Comment (Mazel-Tovr) [inline src/main/kotlin/Core.kt:191]: swap
```suggestion
    if (ids.isEmpty()) return@transaction entities
    val entities: MutableList<T> = mutableListOf()
```
or move this check before creating transaction

File: src/main/kotlin/Core.kt [hunk 12]
```diff
@@ -214,30 +249,93 @@ inline fun <reified T : Any> Transaction.storeAsString(
...
+                }
+                statement.executeBatch()
+            }
+        }
+    } finally {
+        file.delete()
+    }
+}
+
+fun readerToString(value: Reader, maxLength: Int): String? {
```
Comment (LikhanovMaxim) [inline src/main/kotlin/Core.kt:307]: add link to copy method 
move to Util

File: src/main/kotlin/Core.kt [hunk 13]
```diff
@@ -214,30 +249,75 @@ inline fun <reified T : Any> Transaction.storeAsString(
...
             |ON CONFLICT (id) DO UPDATE SET JSON_BODY = excluded.JSON_BODY
         """.trimMargin()
     val stm = connection.prepareStatement(stmt, false)
-    stm[1] = json.encodeToString(T::class.serializer(), any)
+    stm[1] = json.encodeToString(T::class.dsmSerializer(classLoader), any)
     stm.executeUpdate()
 }
 
+inline fun <reified T : Any> storeCollection(
+    any: Iterable<T>,
```
Comment (ViktoriiaLebedeva) [inline src/main/kotlin/Core.kt:265]: rename

File: src/main/kotlin/Utils.kt [hunk 14]
```diff
@@ -15,27 +15,63 @@
...
 
-fun Any.encodeId(): String = this as? String ?: (if(this is Enum<*>) this.toString() else null)?: json.encodeToString(unchecked(this::class.serializer()), this)
+fun <T : Any> KClass<T>.dsmSerializer(
+    classLoader: ClassLoader
+): KSerializer<T> = DsmSerializer(this.serializer(), classLoader)
 
 @Suppress("UNCHECKED_CAST")
 fun <T> unchecked(any: Any) = any as T
+
+fun readerToString(value: Reader, maxLength: Int): String? {
```
Comment (ViktoriiaLebedeva) [inline src/main/kotlin/Utils.kt:61]: Add test with reading and writing UTF symbols

File: src/main/kotlin/serializer/Serializers.kt [hunk 15]
```diff
@@ -67,3 +53,137 @@ object BinarySerializer : KSerializer<ByteArray> {
...
+ */
+class DsmSerializer<T>(
+    private val serializer: KSerializer<T>,
+    val classLoader: ClassLoader
+) : KSerializer<T> by serializer {
+
+    private var parentId: Int? = null
+
+    override fun serialize(encoder: Encoder, value: T) {
+        if (value is BitSet) {
```
Comment (ViktoriiaLebedeva) [inline src/main/kotlin/serializer/Serializers.kt:70]: Try to move it in encodeSerializableElement. Add javadoc

File: src/main/kotlin/serializer/Serializers.kt [hunk 16]
```diff
@@ -67,3 +53,137 @@ object BinarySerializer : KSerializer<ByteArray> {
...
+
+    override fun deserialize(decoder: Decoder): T {
+        if (serializer.descriptor.serialName == BitSet::class.simpleName) {
+            val decodeSerializableValue = decoder.decodeSerializableValue(String.serializer())
+            @Suppress("UNCHECKED_CAST")
+            return decodeSerializableValue.toBitSet() as T
+        }
+        return serializer.deserialize(DsmDecoder(decoder))
+    }
+
```
Comment (Mazel-Tovr) [inline src/main/kotlin/serializer/Serializers.kt:86]: What do you think ? 
```suggestion
    override fun serialize(encoder: Encoder, value: T) {
        if (serializer.isBitSetSerializer()) {
            encoder.encodeSerializableValue(String.serializer(), value.stringRepresentation())
        } else {
            serializer.serialize(DsmEncoder(encoder), value)
        }
    }

    override fun deserialize(decoder: Decoder): T {
        if (serializer.isBitSetSerializer()) {
            val decodeSerializableValue = decoder.decodeSerializableValue(String.serializer())
            @Suppress("UNCHECKED_CAST")
            return decodeSerializableValue.toBitSet() as T
        }
        return serializer.deserialize(DsmDecoder(decoder))
    }
  private fun KSerializer<*>.isBitSetSerializer() = serializer.descriptor.serialName == BitSet::class.simpleName
```

Comment (ViktoriiaLebedeva) [reply src/main/kotlin/serializer/Serializers.kt:86]: Then we have to do cast, but your variant looks better anyway.
`encoder.encodeSerializableValue(String.serializer(), (value as BitSet).stringRepresentation())`

**Pull request merged**